skxray/cdi.py

```diff
@@ -119,9 +125,339 @@ def convolution(array1, array2):
     `this issue on github <https://github.com/Nikea/scikit-xray/issues/258>`_
     for details.
     """
-    fft_norm = lambda x:  np.fft.fftshift(np.fft.fftn(x)) / np.sqrt(np.size(x))
-    fft_1 = fft_norm(array1)
-    fft_2 = fft_norm(array2)
-    return np.abs(np.fft.ifftshift(np.fft.ifftn(fft_1*fft_2)) *
-                  np.sqrt(np.size(array2)))
+    fft_1 = _fft_helper(array1) / np.sqrt(np.size(array1))
+    fft_2 = _fft_helper(array2) / np.sqrt(np.size(array2))
+    return np.abs(_ifft_helper(fft_1*fft_2) * np.sqrt(np.size(array2)))
+
+
+def pi_modulus(array_in, diff_array,
+               thresh_v=1e-12):
+    """
+    Transfer sample from real space to q space.
+    Use constraint based on diffraction pattern from experiments.
+
+    Parameters
+    ----------
+    array_in : array
+        reconstructed pattern in real space
+    diff_array : array
+        experimental data
+    thresh_v : float
+        add small value to avoid the case of dividing something by zero
+
+    Returns
+    -------
+    arr : array
+        updated pattern in q space
+    """
+    diff_tmp = _fft_helper(array_in) / np.sqrt(np.size(array_in))
+    index = np.where(diff_array > 0)
```

Comment:
You can do just `index = diff_array > 0`

Exactly which will be faster will depend on the typical size of these arrays.  from rough testing, dropping the `where` saves 10x, but makes the indexing later about 2x slower.